src/nemos/solvers.py

```diff
@@ -0,0 +1,726 @@
+        tol: float = 1e-3,
+        batch_size: int = 1,
+    ):
+        self.fun = fun
+        self.maxiter = maxiter
+        self.key = key
+        self.stepsize = stepsize
+        self.tol = tol
+        self.loss_gradient = jit(grad(self.fun))
+        self.batch_size = batch_size
+        self.proximal_operator = prox
+
+    def init_state(
+        self,
+        init_params: Any,
+        hyperparams_prox: Any,
+        *args,
+        init_full_gradient: bool = False,
+    ) -> SVRGState:
+        """
+        Initialize the solver state
+
+        Parameters
+        ----------
+        init_params : Any
+            pytree containing the initial parameters.
+            For GLMs it's a tuple of (W, b)
+        hyperparams_prox : float
+            Parameters of the proximal operator, in our case the regularization strength.
+            Not used here, but required to be consistent with the jaxopt API.
+        args:
+            Positional arguments passed to loss function `fun` and its gradient.
+            For GLMs it is:
+                X :
+                    Input data.
+                y :
+                    Output data.
+        init_full_gradient : bool, default False
+            Whether to calculate the full gradient at the initial parameters,
+            assuming that args hold the full data set, and store this gradient in the initial state.
+
+        Returns
+        -------
+        state : SVRGState
+            Initialized optimizer state
+        """
+        df_xs = None
+        if init_full_gradient:
+            df_xs = self.loss_gradient(init_params, *args)
+
+        state = SVRGState(
+            iter_num=0,
+            key=self.key if self.key is not None else random.key(0),
+            error=jnp.inf,
+            stepsize=self.stepsize,
+            # loss_log=jnp.zeros((self.maxiter,)),
+            xs=init_params,
+            df_xs=df_xs,
+            x_av=init_params,
+        )
+        return state
+
+    def _update_loss_log(
+        self,
+        loss_log: jnp.ndarray,
+        i: int,
+        params: Any,
+        *args,
+    ) -> jnp.ndarray:
+        """
+        Update an entry in the array used for storing the log of the loss throughout the optimization.
+
+        Parameters
+        ----------
+        loss_log : jnp.ndarray
+            1D array storing the loss values.
+        i : int
+            Index at which to update, most likely the current iteration number.
+        params : Any
+            Parameters with which to evaluate the loss.
+        args:
+            Positional arguments passed to loss function `fun` and its gradient.
+            For GLMs it is:
+                X :
+                    Input data.
+                y :
+                    Output data.
+
+        Returns
+        -------
+        Updated loss log.
+        """
+        return loss_log.at[i].set(self.fun(params, *args))
+
+    @partial(jit, static_argnums=(0,))
+    def _xk_update(
```

Comment:
I renamed this to `_xk_update_step` to signal that it's a single gradient step. I think I'm leaning towards leaving it like this instead of saying parameter because it's more explicit which step of the SVRG algorithm it refers to, whereas parameter step/update is a bit more vague to me. I feel like it might also discourage misuse or confusion with `update`.

Reply:
I renamed it to `_inner_loop_param_update_step`